### ISSUE TITLE
Better Error Message for SymIntArrayRef expected to contain only concrete integers

### ISSUE THREAD

**Issue by eellison** [pytorch/pytorch#140960]
### 🐛 Describe the bug

Recently it took multiple people and many suggestions to debug why we were getting `SymIntArrayRef expected to contain only concrete integers` error. 

The common reasons are:

- FakeTensorMode not enabled
- Op doesn't support symints
- Pydispatcher not enabled - related to above ^.

We should augment the error message to be more informative and potentially diagnose which of the above it is.

cc @youkaichao 

### Versions

master

**Comment by youkaichao** [comment]
glad to see it, thanks for helping!

**Comment by egaznep** [comment]
I just faced this error, while trying to `materialize` an `UninitializedBuffer` with a dynamic shape. So one can get such an error if you want to instantiate a dynamic LazyModule. The function at the top of the Python callstack was `torch.empty(shape, device=device, dtype=dtype)`, where shape was `(s0, 5, 10)`, `s0` being a `torch.SymInt` object. I think including a printout of the `shape` in the error message would help.

**Comment by ezyang** [comment]
So, concretely, you want to catch and rethrow the exception at a higher level, right? The problem with trying to fix this directly is that the C++ call site is generated implicitly via template metaprogramming inserting coercions so there isn't actually a "place" to fix it directly.

**Comment by ezyang** [comment]
That's very inconvenient. The error is thrown from code in c10. I would need to have access to CPython to do that.

**Comment by eellison** [comment]
@ezyang wouldn't you say this falls under "programming model" ? this is basically an inscruptable error with no clear remediation

**Comment by ezyang** [comment]
I agree it is an inscrutable error. I am simply advising what implementation paths I think will be easier to fix it with :)

**Comment by qsh-zh** [comment]
run into the same issue. The python callstack show the error when I try to call `torch.empty(shape).type_as(tensor)`. The error does not give many useful clues.
